### ISSUE TITLE
High CPU consumption can lead to Unobserved exception

### ISSUE THREAD

**Issue by mazurekj** [rabbitmq/rabbitmq-dotnet-client#465]
Hi,

I am using RabbitMq library in a process with a high CPU load (100% on all 4 cores). In a long time usage of my app, there are appearing some unobserved exceptions from ``` TcpClientAdapter.ConnectAsync ```. I've investigated this issue and it looks like that sometimes this line of code: 
```
Dns.GetHostAddressesAsync(host).ConfigureAwait(false);
```
takes more time than expected, and 

``` SocketFrameHandler.ConnectOrFail ``` ends with an TimeoutException, and disposes TcpClientAdapter.

If this happens, after ``` Dns.GetHostAddressesAsync(host) ``` finally ends, next line:
```
ep = TcpClientAdapterHelper.GetMatchingHost(adds, sock.AddressFamily);
```
throws an unobserved exception, because the "sock" property is already null. 

Original exception text is : 
```
System.AggregateException: Wyj�tk�w zadania nie zaobserwowano ani przez oczekiwanie na zadanie, ani przez uzyskanie dost�pu do jego w�a�ciwo�ci Exception. W wyniku tego niezaobserwowany wyj�tek zosta� wywo�any ponownie przez w�tek finalizatora. ---> System.NullReferenceException: Odwo�anie do obiektu nie zosta�o ustawione na wyst�pienie obiektu.
   w RabbitMQ.Client.TcpClientAdapter.<ConnectAsync>d__2.MoveNext()
   --- Koniec �ladu stosu wyj�tk�w wewn�trznych ---
---> (Wyj�tek wewn�trzny #0) System.NullReferenceException: Odwo�anie do obiektu nie zosta�o ustawione na wyst�pienie obiektu.
   w RabbitMQ.Client.TcpClientAdapter.<ConnectAsync>d__2.MoveNext()<---
```
It's in Polish, but I assume NullReferenceException tells you everything.

Maybe a simple Dispose Pattern could help there to avoid this exceptions.

I'm using the latest 5.1.0 version. 
Latest Windows 10 version.
Erlang OTP 21,
RabbitMQ Server : 3.7.7

**Comment by michaelklishin** [comment]
Thank you for your time.

Team RabbitMQ uses GitHub issues for specific actionable items engineers can work on. GitHub issues are not used for questions, investigations, root cause analysis, discussions of potential issues, etc (as defined by this team).

We get at least a dozen of questions through various venues every single day, often light on details.
At that rate GitHub issues can very quickly turn into a something impossible to navigate and make sense of even for our team. Because GitHub is a tool our team uses heavily nearly every day, the signal/noise ratio of issues is something we care about a lot.

Please post this to [rabbitmq-users](https://groups.google.com/forum/#!forum/rabbitmq-users).

Thank you.

**Comment by michaelklishin** [comment]
See #439, #440 and in particular #433.

**Comment by michaelklishin** [comment]
If you believe this is not a duplicate of #433 and/or not addressed sufficiently by #440, we'd welcome a PR with further improvements. Testing this particular scenario can be pretty hard but tools like Toxiproxy can help emulate some scenarios that stem from runtime resource exhaustion.

**Comment by mazurekj** [comment]
Yes, it's a duplicate.
